File: 13.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func nextDeparture(time int, buses []int) (int, int) {
	var bus int
	tMin := math.MaxInt64
	for _, b := range buses {
		t := b * (time/b + 1)
		if t < tMin {
			tMin = t
			bus = b
		}
	}

	return bus, tMin
}

type Entry struct {
	offset int
	bus    int
}

func matchSeries(series []Entry) int {
	sort.Slice(series, func(i, j int) bool {
		return series[i].offset > series[j].offset
	})
	t := 0
	step := 1
	for _, x := range series {
		for (t+x.offset)%x.bus != 0 {
			t += step
		}
		step *= x.bus
	}
	return t
}

func main() {
	bytes, err := ioutil.ReadFile("13.in")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	lines := strings.Split(strings.TrimSpace(string(bytes)), "\n")

	// Part 1
	time, _ := strconv.Atoi(lines[0])
	buses := []int{}
	for _, x := range strings.Split(lines[1], ",") {
		if i, e := strconv.Atoi(x); e == nil {
			buses = append(buses, i)
		}
	}
	bus, departure := nextDeparture(time, buses)
	fmt.Println((departure - time) * bus)

	// Part 2
	series := []Entry{}
	for offset, s := range strings.Split(lines[1], ",") {
		if x, e := strconv.Atoi(s); e == nil {
			series = append(series, Entry{offset, x})
		}
	}
	fmt.Println(matchSeries(series))
}

File: 13.py
import math
import itertools


def next_departure(time, buses):
    bus = None
    t_min = float("inf")
    for b in buses:
        t = b * math.ceil(time / b)
        if t < t_min:
            t_min = t
            bus = b

    return bus, t_min


def match_series(series):
    series.sort(key=lambda x: x[0], reverse=True)
    t = 0
    step = 1
    for offset, bus in series:
        while (t + offset) % bus:
            t += step
        step *= bus

    return t


if __name__ == "__main__":
    with open("13.in") as f:
        input = [x.strip() for x in f.readlines()]

    # Part 1
    time = int(input[0])
    buses = []
    for x in input[1].split(","):
        if x == "x":
            continue
        buses.append(int(x))

    bus, departure = next_departure(time, buses)
    print((departure - time) * bus)

    # Part 2
    series = []
    for i, x in enumerate(input[1].split(",")):
        if x == "x":
            continue
        else:
            x = int(x)
        series.append((i, x))

    print(match_series(series))
